[RFC] vim-patch:7.4.2220,7.4.2265

jbradaric opened the pull request:
```
vim-patch:7.4.2220

Problem:    printf() gives an error when the argument for %s is not a string.
            (Ozaki Kiichi)
Solution:   Behave like invoking string() on the argument. (Ken Takata)
```
https://github.com/vim/vim/commit/e5a8f35b4286135f3469f3b00a6c2220553d9658

Commit:
vim-patch:7.4.2220

Problem:    printf() gives an error when the argument for %s is not a string.
            (Ozaki Kiichi)
Solution:   Behave like invoking string() on the argument. (Ken Takata)

https://github.com/vim/vim/commit/e5a8f35b4286135f3469f3b00a6c2220553d9658

Shougo commented:
I think 7.4.2221 is also needed.

jbradaric commented:
@Shougo It seems that 7.4.2221 was implemented as a part of #3916

justinmk commented:
@ZyX-I can you look at this?

ZyX-I commented on src/nvim/strings.c:
```diff
@@ -588,18 +589,30 @@ static varnumber_T tv_nr(typval_T *tvs, int *idxp)
 /// @return String value or NULL in case of error.
-static char *tv_str(typval_T *tvs, int *idxp)
-  FUNC_ATTR_NONNULL_ALL FUNC_ATTR_WARN_UNUSED_RESULT
+static char *tv_str(typval_T *tvs, int *idxp, char_u **tofree)
```
Better use `char **const`: we prefer strings to have `char` type.

ZyX-I commented on src/nvim/strings.c:
```diff
@@ -588,18 +589,30 @@ static varnumber_T tv_nr(typval_T *tvs, int *idxp)
-static char *tv_str(typval_T *tvs, int *idxp)
-  FUNC_ATTR_NONNULL_ALL FUNC_ATTR_WARN_UNUSED_RESULT
+static char *tv_str(typval_T *tvs, int *idxp, char_u **tofree)
+  FUNC_ATTR_NONNULL_ARG(1) FUNC_ATTR_NONNULL_ARG(2)
```
Keep FUNC_ATTR_NONNULL_ALL. Also these two may be merged (`FUNC_ATTR_NONNULL_ARG(1, 2)`), but this is not needed: see next comment.

ZyX-I commented on src/nvim/strings.c:
```diff
@@ -588,18 +589,30 @@ static varnumber_T tv_nr(typval_T *tvs, int *idxp)
   } else {
     (*idxp)++;
-    s = (char *)get_tv_string_chk(&tvs[idx]);
+    if (tofree != NULL) {
```
Remove this check and `else` branch (and also modify documenation): it is overengineering. `tv_str` function is used in one place only and where it is used it receives non-NULL argument.

ZyX-I replied:
Though no, do not remove `else` branch. Just use `encode_tv2echo` if `tvs[idx].v_type` is not `VAR_STRING` or `VAR_NUMBER`: this will avoid extra allocation. I would also suggest to set `*tofree` to `NULL` there to be explicit.

ZyX-I commented on src/nvim/strings.c:
```diff
@@ -813,6 +826,9 @@ int vim_vsnprintf(char *str, size_t str_m, const char *fmt, va_list ap,
       char fmt_spec = '\0';
 
+      // buffer for 's' and 'S' specs
+      char_u *tofree = NULL;
```
`char *`.

ZyX-I commented on src/nvim/strings.c:
```diff
@@ -588,18 +589,30 @@ static varnumber_T tv_nr(typval_T *tvs, int *idxp)
 ///                  value.
 /// @param[in,out]  idxp  Index in a list. Will be incremented.
 ///
+/// @param[out]  tofree If "tofree" is NULL get_tv_string_chk() is used. Some
```
Should be two spaces after the name of the variable. Also modify documentation, it may not be NULL in the code.

Commit:
strings.c: Fix problems found during code review.

Commit:
vim-patch:7.4.2265

Problem:    printf() isn't tested much.
Solution:   Add more tests for printf(). (Dominique Pelle)

https://github.com/vim/vim/commit/76efafba2af36ae5f6c7b79b56c537fcbcdb386c

jbradaric commented:
@ZyX-I I've fixed the problems you've found and also added more tests for `printf` (patch 7.4.2265).

ZyX-I reviewed (approved):
Good, except that constant pointers are normally written like `**const` (no space before `const`).

jbradaric commented:
@ZyX-I I've also found some other patches with fixes for printf (7.4.2266, 7.4.2280, 7.4.2291), I'll add those as well.
